File: flutter/amplify_and_flutter_proto/lib/auth/auth_service.dart
import 'dart:async';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_core/amplify_core.dart';
import 'package:flutter/material.dart';
import 'auth_credentials.dart';
import 'auth_flow_status.dart';
import 'sign_up_credentials.dart';

class AuthState {
  AuthState({@required this.authFlowStatus, this.user});

  final AuthFlowStatus authFlowStatus;
  final AuthUser user;
}

class AuthService {
  AuthCredentials _credentials;
  final authStateController = StreamController<AuthState>();

  void getCurrentUser() async {
    try {
      final user = await Amplify.Auth.getCurrentUser();
      final state =
          AuthState(authFlowStatus: AuthFlowStatus.session, user: user);

      authStateController.add(state);

      _credentials = null;
    } catch (_) {
      print('user not logged in');
      showLogin();
    }
  }

  void showSignUp() {
    final state = AuthState(authFlowStatus: AuthFlowStatus.signUp);
    authStateController.add(state);
  }

  void showLogin() {
    final state = AuthState(authFlowStatus: AuthFlowStatus.login);
    authStateController.add(state);
  }

  void signUpWithCredentials(SignUpCredentials credentials) async {
    try {
      final Map<String, dynamic> userAttributes = {'email': credentials.email};

      final result = await Amplify.Auth.signUp(
          username: credentials.username,
          password: credentials.password,
          options: CognitoSignUpOptions(userAttributes: userAttributes));

      if (result.isSignUpComplete) {
        loginWithCredentials(credentials);
      } else {
        this._credentials = credentials;

        final state = AuthState(authFlowStatus: AuthFlowStatus.verification);
        authStateController.add(state);
      }
    } on AuthError catch (e) {
      print('Sign up error - ${e.cause}');
    }
  }

  void verifyConfirmation(String code) async {
    if (this._credentials == null) return;

    try {
      final result = await Amplify.Auth.confirmSignUp(
          username: this._credentials.username, confirmationCode: code);

      if (result.isSignUpComplete) {
        loginWithCredentials(_credentials);
      } else {
        // Follow more sign up steps
      }
    } catch (e) {
      print(e);
    }
  }

  void loginWithCredentials(AuthCredentials credentials) async {
    try {
      final result = await Amplify.Auth.signIn(
          username: credentials.username, password: credentials.password);

      if (result.isSignedIn) {
        getCurrentUser();
      } else {
        print('user is not signed in');
      }
    } on AuthError catch (e) {
      print('could not log in - ${e.cause}');
    }
  }

  void logOut() async {
    try {
      final options = CognitoSignOutOptions(globalSignOut: true);
      await Amplify.Auth.signOut(options: options);

      final state = AuthState(authFlowStatus: AuthFlowStatus.login);
      authStateController.add(state);

      print('signed out');
    } catch (e) {
      print('Log out error - $e');
    }
  }
}

File: apple/hotel-booking-proto/hotel-booking-proto/Auth/AuthService.swift
//
//  AuthService.swift
//  hotel-booking-proto
//
//  Created by Kilo Loco on 11/4/20.
//

import Amplify
import Foundation

enum AuthState {
    case login, signUp, confirmation, session(guest: Guest)
}

class AuthService: ObservableObject {
    
    @Published var state: AuthState = .login
    
    private var authCredentials: AuthCredentials?
    
    func getCurrentUser() {
        if let currentUser = Amplify.Auth.getCurrentUser() {
            print("Signed in", currentUser)
            let guest = Guest(
                id: currentUser.userId,
                username: currentUser.username
            )
            DispatchQueue.main.async { [weak self] in
                self?.state = .session(guest: guest)
            }
            
        } else {
            print("Not signed in")
            showLogin()
        }
    }
    
    func showSignUp() {
        DispatchQueue.main.async { [weak self] in
            self?.state = .signUp
        }
    }
    
    func showLogin() {
        DispatchQueue.main.async { [weak self] in
            self?.state = .login
        }
    }
    
    func signUp(with credentials: SignUpCredentials) {
        self.authCredentials = credentials
        
        let emailAttribute = AuthUserAttribute(.email, value: credentials.email)
        let usernameAttribute = AuthUserAttribute(.preferredUsername, value: credentials.username)
        let options = AuthSignUpRequest.Options(
            userAttributes: [emailAttribute, usernameAttribute]
        )
        
        Amplify.Auth.signUp(
            username: credentials.username,
            password: credentials.password,
            options: options
        ) { [weak self] (result) in
            do {
                let nextStep = try result.get().nextStep
                
                switch nextStep {
                case .confirmUser:
                    DispatchQueue.main.async {
                        self?.state = .confirmation
                    }
                    
                case .done:
                    self?.login(with: credentials)
                }
                
            } catch {
                print("Sign up error:", error)
            }
        }
    }
    
    func confirmEmail(with confirmationCode: String) {
        guard let credentials = self.authCredentials else { return }
        
        Amplify.Auth.confirmSignUp(
            for: credentials.username,
            confirmationCode: confirmationCode
        ) { [weak self] (result) in
            do {
                let nextStep = try result.get().nextStep
                
                switch nextStep {
                case .done:
                    self?.login(with: credentials)
                    
                default:
                    print("This should never happen but the next step is:", nextStep)
                }
                
            } catch {
                print("Could not confirm", error)
            }
        }
    }
    
    func login(with credentials: AuthCredentials) {
        
        Amplify.Auth.signIn(username: credentials.username, password: credentials.password) { [weak self] (result) in
            do {
                
                let isSignedIn = try result.get().isSignedIn
                
                if isSignedIn {
                    self?.authCredentials = nil
                    self?.getCurrentUser()
                    
                } else {
                    print("Is not signed in but sign in was successful?!?!?!?")
                }
                
            } catch {
                print("Could not sign in:", error)
            }
        }
        
        
    }
    
    func logOut() {
        _ = Amplify.Auth.signOut()
        showLogin()
    }
    
}
